### ISSUE TITLE
Alternative trie dependency?

### ISSUE THREAD

**Issue by goodmami** [LuminosoInsight/langcodes#35]
This project depends on [marisa-trie](https://github.com/pytries/marisa-trie), which is a Cython/C++ extension module. The [latest release](https://pypi.org/project/marisa-trie/0.7.5/#files) from 2.5 years ago only has pre-built binaries for Python3.6 on macOS, and there are no releases with binaries for a newer version of Python than 3.6. This means users must build it from source, which requires them to have a compiler, the Python development headers, etc. installed. I created [this issue](https://github.com/pytries/marisa-trie/issues/57) but haven't gotten a response from the marisa-trie devs.

Would you be receptive to using a more active project, like [datrie](https://github.com/pytries/datrie), or a pure-Python one like [pygtrie](https://github.com/mina86/pygtrie)?

I'm wary to distribute software that depends on langcodes because of the problems with the marisa-trie dependency.

**Comment by ygorg** [comment]
First of all thank you so much this awesome library, you make working with languages as easy as it should be !! 
I upvote this issue, in the past few weeks there were user having trouble with installing `marisa-trie` on macOS.

I did some benchmarking by using `dict` and using `pygtrie` (as recommended), in order to use those implentations I put the data in json dict format, I also compressed the whole `data/trie` folder using `tar -xvzf`, which compressed pretty well.
I only used `example.py` for benchmarking which is not enough to test everything.

But as a general thing not using marisa-trie is slower and heavier.
Below are my results and the some useful code to reproduce.


|  | marisa-trie | python dict | pygtrie |
|:----- | ---------------:| --------------:| ----:|
| example.py (s) | 1.5 | 19.5 | 5.8 |
| size of data/trie (Mo) | 7.7 | 29.3 | 29.3 |
| example.py gz (s) | - | 20.2 | 6.3 |
| size of trie.tar.gz (Mo) | - | 4.9 | 4.9 |

```python
# Convert Tries to json dicts
import marisa_trie, json
from glob import glob
for fp in glob('trie/*.marisa')+glob('trie/*/*.marisa'):
    trie = marisa_trie.BytesTrie()
    trie.load(fp)
    dd = {k: v.decode('utf-8') if type(v) is bytes else v for k, v in trie.items()}
    with open(fp.replace('marisa', 'json'), 'w', encoding='utf-8') as f:
        json.dump(dd, f)

# names.py
# Load pygtrie
def load_trie(filename):
    trie = pygtrie.CharTrie()
    with open(filename, encoding='utf-8') as f:
        for k, v in json.load(f).items():
            trie[k] = v
    return trie

# Load 
def load_trie(filename):
    data_path = os.path.abspath(os.path.join(os.path.dirname(__file__), 'data'))
    filename = filename.replace(os.path.commonpath((data_path, filename)), '')
    filename = filename[1:] if filename[0] == os.sep else filename
    filename = os.path.join(*(filename.split(os.sep)[1:]))
    trie = pygtrie.CharTrie()
    with tarfile.open(os.path.join(data_path, 'trie.tar.gz'), 'r:gz', encoding='utf-8') as g:
        f = g.extractfile(filename)
        for k, v in json.load(f).items():
            trie[k] = v
    return trie

# marisa_trie.Trie.keys(lookup)
possible_keys = [k for k in trie if k.startswith(lookup)]  # dict
possible_keys = trie.keys(lookup)  # pygtrie
# marisa_trie.Trie.prefixes(lookup)
prefixes = [lookup[:i] for i in range(len(lookup)) if lookup[:i] in trie]  # both dict and pygtrie
```

**Comment by goodmami** [comment]
Thanks for the comparisons, @ygorg! I wonder if speed is crucial for any users of this library, because I think a 4x slowdown, while not great, is an acceptable tradeoff to increase portability.

Also, I just want to add that when I looked into datrie, it seems that it doesn't have full support for unicode, which might be an issue here.

**Comment by 2press** [comment]
Due to the dependence on marisa-trie the package cannot be installed on Python 3.9 :(

**Comment by ygorg** [comment]
@2press could you elaborate on why this happens ?

**Comment by 2press** [comment]
I tested his PR but couldn't build it either.

**Comment by vincentbernat** [comment]
Most users installing with `pip` can't use the hack.

**Comment by goodmami** [comment]
@ygorg I adapted the code you wrote above into the linked PR. Instead of tarring up the whole `data/trie/` subdirectory, I gzip-compressed each json file. The performance was similar to what you reported above. Thanks again for sharing!

**Comment by goodmami** [comment]
@rspeer sorry to poke, but do you have a minute to comment on this or the related PR #35? It's been sitting here a while and some projects (see above) have already moved on.

**Comment by Tahnan** [comment]
@goodmami - one of Robyn's coworkers here.  She's on a well-earned, much-deserved sabbatical for the last three months of the year; she might be checking in on things, but there's a good chance she hasn't been looking at work-related notifications at all, and won't get to it until she's back in January.  Just wanted to reassure you that she's not forgetting or ignoring it!

**Comment by goodmami** [comment]
@Tahnan Got it. Thanks for the context!

**Comment by nyuszika7h** [comment]
`marisa-trie-m` can be used as a drop-in replacement on Python 3.9. People depending on `langcodes` can put this in their `requirements.txt` file for now:

```
marisa-trie-m
--install-option="--no-deps" langcodes
```

**Comment by rspeer** [comment]
I implemented #37 as a short-term fix, and released version 2.2. I'm going to look into #14 as a better medium-term fix.
